From 11af775626d3fcdfa70808fd83bde058a92c10c8 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:48:22 -0400
Subject: [x86] perf: Correctly use FEATURE_PDCM

Message-id: <1408953210-26343-15-git-send-email-jolsa@redhat.com>
Patchwork-id: 88734
O-Subject: [PATCH RHEL7.1 BZ1133083 014/322] perf/x86: Correctly use FEATURE_PDCM
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit c9b08884c9c98929ec2d8abafd78e89062d01ee7
Author: Peter Zijlstra <peterz@infradead.org>
Date: Mon Feb 3 14:29:03 2014 +0100

description
===========
The current code simply assumes Intel Arch PerfMon v2+ to have
the IA32_PERF_CAPABILITIES MSR; the SDM specifies that we should check
CPUID[1].ECX[15] (aka, FEATURE_PDCM) instead.

This was found by KVM which implements v2+ but didn't provide the
capabilities MSR. Change the code to DTRT; KVM will also implement the
MSR and return 0.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/cpu/perf_event_intel.c b/arch/x86/kernel/cpu/perf_event_intel.c
index 1c395ee..40dc8c6 100644
--- a/arch/x86/kernel/cpu/perf_event_intel.c
+++ b/arch/x86/kernel/cpu/perf_event_intel.c
@@ -2303,10 +2303,7 @@ __init int intel_pmu_init(void)
  if (version > 1)
   x86_pmu.num_counters_fixed = max((int)edx.split.num_counters_fixed, 3);
 
- /*
-  * v2 and above have a perf capabilities MSR
-  */
- if (version > 1) {
+ if (boot_cpu_has(X86_FEATURE_PDCM)) {
   u64 capabilities;
 
   rdmsrl(MSR_IA32_PERF_CAPABILITIES, capabilities);
-- 
1.7.1